### PR TITLE
added API for suggested users'

### PR DESCRIPTION
**Pull request by vinayak-trivedi**
- Added API for giving suggested users according a particular school

### CODE REVIEW

File: models/users.js [hunk 1]
```diff
@@ -83,6 +84,35 @@ const getJoinData = async (userId) => {
...
+        usersId.push(doc.data().itemid);
+      });
+    }
+    const Promises = usersId.map((userId) => fetchUser({ userId }));
+    const resolved = await Promise.all(Promises);
+    const set = new Set();
+    const users = [];
+
+    resolved.forEach((resolve) => {
+      if (!set.has(resolve.user.id)) {
```
Comment (DashDeipayan) [inline models/users.js:103]: can you please add a bit more descriptive names to the variable names of `Promises`, `resolved`, `set`, `resolve`

File: routes/users.js [hunk 2]
```diff
@@ -23,5 +23,6 @@ router.post("/picture", authenticate, upload.single("profile"), users.postUserPi
 router.patch("/profileURL", authenticate, userValidator.updateProfileURL, users.profileURL);
 router.patch("/rejectDiff", authenticate, authorizeRoles([SUPERUSER]), users.rejectProfileDiff);
 router.patch("/:userId", authenticate, authorizeRoles([SUPERUSER]), users.updateUser);
+router.get("/suggestedUsers/:skillId", authenticate, authorizeRoles([SUPERUSER]), users.getSuggestedUsers);
```
Comment (DashDeipayan) [inline routes/users.js:26]: NIT comment: It would be better if you could de-structure all the controller functions of `users` instead of using `users.*`

**Review by DashDeipayan — CHANGES_REQUESTED** [review]

**Review by ivinayakg — APPROVED** [review]
LGTM
Thank you for making the suggested changes.

**Review by DashDeipayan — DISMISSED** [review]
LGTM

File: models/users.js [hunk 3]
```diff
@@ -83,6 +84,42 @@ const getJoinData = async (userId) => {
...
+    const data = await itemModel.where("itemtype", "==", "USER").where("tagid", "==", skill).get();
+    const usersId = [];
+    const users = [];
+
+    if (!data.empty) {
+      data.forEach((doc) => {
+        usersId.push(doc.data().itemid);
+      });
+      const usersArray = usersId.map((userId) => fetchUser({ userId }));
+      const userDuplicate = await Promise.all(usersArray);
```
Comment (ankushdharkar) [inline models/users.js:105]: I am not sure if the naming here looks right. What does `userDuplicate` signify? It looks like the `usersArray` is just used to get back a list of users' data.

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: models/users.js [hunk 4]
```diff
@@ -84,6 +84,39 @@ const getJoinData = async (userId) => {
...
+          usersId.push(doc.data().itemid);
+        }
+      });
+      const usersArray = usersId.map((userId) => fetchUser({ userId }));
+      users = await Promise.all(usersArray);
+    }
+
+    return { users };
+  } catch (err) {
+    logger.error("Error in getting suggeted", err);
```
Comment (ankushdharkar) [inline models/users.js:115]: Typo

Comment (ankushdharkar) [reply models/users.js:115]: @vinayak-trivedi Please fix this in a future feature :D

File: models/users.js [hunk 5]
```diff
@@ -84,6 +84,39 @@ const getJoinData = async (userId) => {
...
+    let users = [];
+
+    const dataSet = new Set();
+
+    if (!data.empty) {
+      data.forEach((doc) => {
+        if (!dataSet.has(doc.data().itemid)) {
+          dataSet.add(doc.data().itemid);
+          usersId.push(doc.data().itemid);
+        }
```
Comment (ankushdharkar) [inline models/users.js:107]:
```suggestion
        const docUserId = doc.data().itemid;
        if (!dataSet.has(docUserId)) {
          dataSet.add(docUserId);
          usersId.push(docUserId);
        }
```

File: models/users.js [hunk 6]
```diff
@@ -84,6 +84,39 @@ const getJoinData = async (userId) => {
...
+    const usersId = [];
+    let users = [];
+
+    const dataSet = new Set();
+
+    if (!data.empty) {
+      data.forEach((doc) => {
+        if (!dataSet.has(doc.data().itemid)) {
+          dataSet.add(doc.data().itemid);
+          usersId.push(doc.data().itemid);
```
Comment (ankushdharkar) [inline models/users.js:106]: Why do we need to maintain two collections? Can we not proceed just with one, and convert it to an array later?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: models/users.js (hunk 4 above)
Comment (vinayak-trivedi) [reply models/users.js]: fixed it, thanks

**Review by ankushdharkar — APPROVED** [review]

**Pull request merged**